janusgraph-cql/src/main/java/org/janusgraph/diskstorage/cql/CQLKeyColumnValueStore.java

```diff
@@ -434,43 +443,48 @@ public KeyIterator getKeys(final SliceQuery query, final StoreTransaction txh) t
     private class CQLPagingIterator implements Iterator<Row> {
 
         private ResultSet currentResultSet;
+        private java.util.Iterator<Row> currentPageIterator;
 
         private int index;
-        private int paginatedResultSize;
-        private final Supplier<Statement> statementSupplier;
+        private final int paginatedResultSize;
 
-        private byte[] lastPagingState = null;
+        private BoundStatement boundStatement;
 
-        public CQLPagingIterator(final int pageSize, Supplier<Statement> statementSupplier) {
-            this.index = 0;
+        public CQLPagingIterator(final int pageSize, BoundStatement boundStatement) {
             this.paginatedResultSize = pageSize;
-            this.statementSupplier = statementSupplier;
-            this.currentResultSet = getResultSet();
+            this.boundStatement = boundStatement;
+            updateResultSet();
         }
 
         @Override
         public boolean hasNext() {
-            return !currentResultSet.isExhausted();
+            if(index<paginatedResultSize){
+                return currentPageIterator.hasNext();
+            }
+            if(currentResultSet.isFullyFetched()){
+                return false;
```

Comment:
Nice! Just a nitpick:
Now this line seems unreachable. See codecov: https://app.codecov.io/gh/JanusGraph/janusgraph/compare/2169/diff

Also sample test code:

```java
    @Test
    public void testPagingIterator() {
        clopen(option(PAGE_SIZE), 10);
        int numVertices = 10;
        for (int i = 0; i < numVertices; i++) {
            graph.addVertex().property("prop", i);
        }
        graph.tx().commit();

        List<Object> values = graph.traversal().V().values("prop").toList();
        assertEquals(numVertices, values.size());
        assertEquals(numVertices, new HashSet<>(values).size());
    }
```

Seems the underlying Datastax driver needs to execute the query once more to know that the current result set is fully fetched. Thus this if-statement seems to be useless.

Reply:
Interestingly, this line is reachable and is necessary when paginatedResultSize is negative. Seems Datastax driver allows negative value, simply ignores it, and uses the default value. If we don't have this if-statement, our program will go into an endless loop.

Reply:
The following test case fails under CQLGraphTest but can pass InMemoryGraphTest & BerkeleyGraphTest:
```java    
    @Test
    public void testPagingIterator() {
        clopen(option(PAGE_SIZE), 10);
        int numVertices = 10;
        for (int i = 0; i < numVertices; i++) {
            graph.addVertex().property("prop", i);
        }
        graph.tx().commit();

        List<Object> values = graph.traversal().V().values("prop").toList();
        assertEquals(numVertices, values.size());
        assertEquals(numVertices, new HashSet<>(values).size());
    }
```
didn't check thoroughly but I guess it's because the implementation here assumes paging size is non-negative.

Reply:
Oh, didn't think about this edge case. For some reason I thought that negative value isn't allowed. I will fix that